### PR TITLE
StarWatch Terraform build implementation.

### LINKED ISSUE #66: Add terraform file names to the gitignore
As a developer or DevOps engineer, I want to add specific Terraform file patterns to the .gitignore file, so that I can prevent temporary or sensitive Terraform files from being accidentally committed to version control. 
Files should include at a minimum:

-  .terraform
-  .terraform.lock.hcl
-  terraform.*
-  .terraform.tfstate.lock.info
-  *.env
-  *.pemkey
-  *.pem

### PR DESCRIPTION
**Pull request by MurasatoNaoya**
### The Provision Cloud Resources for the StarWatch Project

This Terraform file will provision the cloud resources needed for our StarWatch project. Below is a high-level description of the services created by `main.tf` and the general functions they serve:

#### 1. Compute Resources
- Uses the existing **C13 ECS cluster**.
- Creates a **task definition** to be run as a dashboard service.
- Configures the **ECS service** to automatically restart in the event of a failure for robustness.
- Sets up multiple **Lambda functions** using ECR images for weekly, hourly, and quad-hourly pipelines.

#### 2. Database Setup
- Creates a **highly available PostgreSQL RDS instance** across two Availability Zones (with the possibility of adding more).
- Defines the necessary **subnet groups** for networking.

#### 3. Security Groups
- Defines **security groups** to control inbound (ingress) and outbound (egress) traffic for both RDS and ECS services.

#### 4. Automated Event-Driven Pipelines
- Implements **EventBridge rules** to trigger data pipelines:
  - Weekly and hourly pipelines are triggered by **Step Functions**.
  - The quad-hourly weather pipeline is triggered directly by **EventBridge**.

#### 5. Parallel Lambda Invocation
- Defines **Step Functions** to run related pipelines in parallel.

#### 6. IAM Roles & Permission Policies
- Creates and attaches **IAM roles and policies** for ECS, Lambda, and Step Functions.
- Ensures functional access for ECR and RDS instances, as well as logging permissions.

#### 7. Logging & Monitoring
- Configures **CloudWatch log groups** to capture detailed execution data for Lambda functions and Step Functions.
- Enables logging for ECS tasks within the **task definition** to monitor container activity and assist in debugging.

Closes #66 #67

### CODE REVIEW

File: terraform/variables.tf
Comment (olliejth) [inline terraform/variables.tf:1]: Could you add the following variables to the relevant files please?
I'll send the corresponding values over slack and update the google doc.

ASTRONOMY_ID
ASTRONOMY_SECRET
DB_HOST
DB_PORT

File: .gitignore [hunk 1]
```diff
@@ -93,4 +93,6 @@ terraform.tfstate.backup
 .pemkey
 *.xml
 logs/
-*.log
\ No newline at end of file
+*.log
+
+
```
Comment (AlexOBarnes) [inline .gitignore:98]: Was this an error?

File: terraform/main.tf [hunk 2]
```diff
@@ -0,0 +1,557 @@
...
+
+# Cohort 13 designated VPC.
+data "aws_vpc" "c13-vpc" {
+  id = var.VPC_ID
+}
+
+
+# Two public subnets for the RDS, to 
+# ensure the database is robust and highly 
+# available across different availablity zones.
```
Comment (AlexOBarnes) [inline terraform/main.tf:19]: Very thorough comments

File: terraform/main.tf [hunk 3]
```diff
@@ -0,0 +1,557 @@
+# Defining provider - AWS.
+# AWS access and secret key are passed in a environment
+# variables from the terraform.tfvars file.
+provider "aws"{
+    region = var.AWS_REGION
+    access_key = var.AWS_ACCESS_KEY 
+    secret_key = var.AWS_SECRET_KEY
```
Comment (AlexOBarnes) [inline terraform/main.tf:7]: Good consideration for security of our AWS secret keys

File: terraform/main.tf [hunk 4]
```diff
@@ -0,0 +1,557 @@
...
+        prevent_destroy = true
+    }
+}
+
+
+# Security group for the ECS service hosting the dashboard.
+# Ingress is set to any traffic on port 8501, crucial for
+# people accessing the dashboard online, as in-bound traffic is blocked by default.
+# Egress is set to any protocol for any IP-address, as the script hosted on
+# the ECS as a service also has to fetch data from external APIs (NASA and ISS location APIs).
```
Comment (AlexOBarnes) [inline terraform/main.tf:60]: Love the comments

File: terraform/main.tf [hunk 5]
```diff
@@ -0,0 +1,557 @@
...
+        prevent_destroy = true
+    }
+}
+
+
+# Defining each individual lambda function for the StarWatch project would lead to 
+# a very long and verbose main.tf file. Instead, a lambda function dictionary has
+# been defined for each Lambda, outlining that name of each pipeline and the 
+# ECR image it references. This way, we can traverse it alone to build lambda functions
+# for each, without having to manually list each configuration out.
```
Comment (AlexOBarnes) [inline terraform/main.tf:131]: Clear comments are great!

File: terraform/main.tf [hunk 6]
```diff
@@ -0,0 +1,557 @@
...
+		invoked_by    = "step_function_hourly"
+    },
+    quadhourly_weather = {
+		function_name = "QuadhourlyOpenMeteoWeatherPipeline"
+		image_uri     = var.QUADHOURLY_WEATHER_IMAGE_URI
+		invoked_by    = "eventbridge_quadhourly"
+    }
+  }
+}
+
```
Comment (AlexOBarnes) [inline terraform/main.tf:161]: Good use of terraform to make code more succinct

File: terraform/main.tf
Comment (olliejth) [inline terraform/main.tf:1]: Really extensive, thanks for this Andrew! Protecting the RDS between terraform destroys was definitely a good idea.

File: terraform/variables.tf [hunk 7]
```diff
@@ -0,0 +1,76 @@
+variable "AWS_REGION" {
+  type = string
```
Comment (AlexOBarnes) [inline terraform/variables.tf:2]: Good considerations for security

**Review by olliejth — CHANGES_REQUESTED** [review]

File: terraform/variables.tf [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+variable "AWS_ACCESS_KEY" {
+  type = string
+  
+}
+
+variable "AWS_SECRET_KEY" {
+  type = string
+  
+}
```
Comment (AlexOBarnes) [inline terraform/variables.tf:15]: We previously had problems with uploading these variables as env variables so keep an eye out for this error in future

File: terraform/main.tf [hunk 9]
```diff
@@ -0,0 +1,557 @@
...
+		security_groups  = [aws_security_group.c13_ecs_service_sg.id]  
+		assign_public_ip = true  # Required for internet access, similar to the toggle in the AWS UI. 
+	}
+}
+
+
+resource "aws_cloudwatch_log_group" "sfn_log_group" {
+  	name = "/aws/states/starwatch_state_machines"
+  	retention_in_days = 14 # Will be deleted after the two week project, 
+						   # if not already terraform destroyed.
```
Comment (AlexOBarnes) [inline terraform/main.tf:556]: Very useful to store our logs

File: terraform/main.tf [hunk 10]
```diff
@@ -0,0 +1,557 @@
...
+	role       = aws_iam_role.sfn_exec_role.name
+	policy_arn = aws_iam_policy.sfn_lambda_permission_policy.arn
+}
+
+
+# This iteratively checks the kind of lambda being assessed, if it's a 
+# lambda that is invoked by a step-function, it is provided permissions allowing
+# that mapped step-function to run it. However, if the lambda being considered is 
+# the quadhourly weather pipeline Lambda, eventbridge is  allowed to trigger it directly,
+# as no step function is involved in the running of that pipeline.
```
Comment (AlexOBarnes) [inline terraform/main.tf:460]: Love the use of this iteration

**Review by AlexOBarnes — APPROVED** [review]
Looks good overall. We should just check the env variable in each pipeline is consistent

File: terraform/variables.tf
Comment (MurasatoNaoya) [reply terraform/variables.tf:1]: I'm not 100% about the question. DB HOST should be in the doc and HB_PORT is default 5432 for Postgres. For the ASTRONOMY_ID and ASTRONOMY_SECRET, I don't see how they'd be used when building services using Terraform?

**Pull request merged**